File: LongestSubstringWithoutRepeatingCharacter.py
# TC: O(N) since we are traversing all the elements in the string. 
# SC: O(N) in worst case, we will end up inserting all the characters of the input string into the hashmap. 

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if not s or len(s) == 0: 
            return 0
        
        slow = fast = max_len = 0
        hmap = {}
        while fast < len(s): 
            curr = s[fast]
            if curr in hmap.keys(): 
                slow = max(slow, hmap.get(curr))
            hmap[curr] = fast + 1
            max_len = max(max_len, (fast - slow) + 1)
            fast += 1
        
        return max_len

File: LongestSubstringWithoutRepeatingCharacter.java
// Time Complexity: O(n)
// Space Complexity: O(1)

/*
 * we keep track of all characters in the string and if we encounter
 * the same character in the string we increment the left pointer until
 * that value is no longer in the hashmap.
 */

class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashSet<Character> map = new HashSet<>();
        int left = 0;
        int maxLength = 0;
        for(int i = 0; i < s.length(); i++){
            if(!map.contains(s.charAt(i))){
                map.add(s.charAt(i));
            }
            else{
                while(s.charAt(left) != s.charAt(i)){
                    map.remove(s.charAt(left));
                    left++;
                }
                left++;
            }
            maxLength = Math.max(maxLength, i-left+1);
        }
        return maxLength;
    }
}